### ISSUE TITLE
React 19 + Cloudflare Adapter -> "Bug: Uncaught ReferenceError: MessageChannel is not defined"

### ISSUE THREAD

**Issue by ethanniser** [withastro/astro#12824]
### Astro Info

```block
Astro                    v5.1.1
Node                     v22.6.0
System                   macOS (arm64)
Package Manager          bun
Output                   server
Adapter                  @astrojs/cloudflare
Integrations             @astrojs/mdx
                         @astrojs/sitemap
                         @astrojs/tailwind
                         @astrojs/react
```

### Describe the Bug

originally shared facebook/react#31827 but I think it makes sense to be an issue here as I doubt its a react issue

![image](https://github.com/user-attachments/assets/98715f11-f094-486c-8d10-5ff108e19555)

It doesn't happen during build but instead during the "Deploying to Cloudflare's global network" step

### What's the expected result?

The error should not happen

### Link to Minimal Reproducible Example

https://github.com/phase/test-astro-react19-cf-workers (thank you @phase)

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by phase** [comment]
here's a workaround https://github.com/withastro/adapters/pull/436#issuecomment-2525190557

`astro.config.mjs`
```typescript
// @ts-check
import { defineConfig } from "astro/config";
import cloudflare from "@astrojs/cloudflare";

import react from "@astrojs/react";

// https://astro.build/config
export default defineConfig({
  output: "server",
  adapter: cloudflare(),
  integrations: [react()],
  vite: {
    resolve: {
      // Use react-dom/server.edge instead of react-dom/server.browser for React 19.
      // Without this, MessageChannel from node:worker_threads needs to be polyfilled.
      alias: import.meta.env.PROD && {
        "react-dom/server": "react-dom/server.edge",
      },
    },
  },
});
```

**Comment by acnebs** [comment]
As it seems unlikely this will be upstream fixed by Cloudflare this probably needs to be addressed on the Astro end.

**Comment by IgorMinar** [comment]
This turns out not to be a Cloudflare issue at all, but rather an odd misconfiguration of of Vite in Astro's Cloudflare plugin.

The culprit appears to be these lines:

https://github.com/withastro/adapters/blob/7b352ee1e52eaa96e6f0dff8be6de573ed7c46c7/packages/cloudflare/src/index.ts#L221-L226

Note how the Astro's adapter here aliases `react-dom/server` to `react-dom/server.browser`, which then contains the offending `MessageChannel` API call which throws because it's not defined.

I'm not sure why Astro created this alias in the first place, but this appears to me to be either obsolete workaround for some no longer present issue, or incorrect workaround for some existing issue. This alias was introduced in the initial commit of the adapter and there are no comments that I can see describing why it exists at all:
- (original repo) https://github.com/withastro/astro/commit/7f423581411648c9a69b68918ff930581f12cf16#diff-2039ec525c2c4d3855197c5d569ef62e843bfb2fe2bc614b206bf541a3d9a46cR36
- (new adapter repo) https://github.com/withastro/adapters/commit/199894c62ae9b350d39ea5f09884f5c2e19a8ee6#diff-1d2a2b56ad20e7e5b13936929d10bebf30d56ffa7cdc64f686dff4ae6cfafb60R36

Changing this alias to point to `react-dom/server.edge` instead of `react-dom/server.browser` is enough to build the application and deploy it to cloudflare: https://a811a373.astro-5-test.pages.dev/

I think it's not right that this alias exist in the first place though, so I dug a but more and tried to remove it which surfaced a new issue related to `react-dom/server` being externalized and not bundled into the worker bundle which then causes a new deployment failure related to `react-dom/server` failing to resolve on Cloudflare.

Digging a bit more, I realized that even though the plugin currently tells vite to disable any externals:
https://github.com/withastro/adapters/blob/7b352ee1e52eaa96e6f0dff8be6de573ed7c46c7/packages/cloudflare/src/index.ts#L244

It then also proceeds to configure `externals` with a code that seems to be mistakingly copy pasted from the Astro Vue adapter:

https://github.com/withastro/adapters/blob/7b352ee1e52eaa96e6f0dff8be6de573ed7c46c7/packages/cloudflare/src/index.ts#L249-L256

This then results in `react-dom/server` to appearing in `externals: [...]` array in the final Vite config:
```
  vite:config     ssr: {
  vite:config       noExternal: true,
  vite:config       external: [ 'react-dom/server', 'react-dom/client' ],
  vite:config       resolve: {
  vite:config         conditions: [
  vite:config           'module',
  vite:config           'browser',
  vite:config           'development|production',
  vite:config           'workerd',
  vite:config           'worker',
  vite:config           'astro',
  vite:config           'astro'
  vite:config         ]
  vite:config       },
  vite:config       target: 'webworker'
  vite:config     },
```

This configuration is invalid, as workerd/Cloudflare doesn't allow for any npm packages to be externalized during bundling. All deps must be inlined.

So the right thing to do here, is to remove the externals config copy-pasted from vue mentioned above, and allow for `react-dom/server` to be bundled along with the application.

This then results again a correct bundle and working astro app.

One last thing to mention, is that since to configure `conditions` we append `workerd` and `worker` to the `conditions` array via push:
https://github.com/withastro/adapters/blob/7b352ee1e52eaa96e6f0dff8be6de573ed7c46c7/packages/cloudflare/src/index.ts#L241

This results in the vite config looking like this:
```
  vite:config     ssr: {
  vite:config       noExternal: true,
  vite:config       external: [ 'react-dom/server', 'react-dom/client' ],
  vite:config       resolve: {
  vite:config         conditions: [
  vite:config           'module',
  vite:config           'browser',
  vite:config           'development|production',
  vite:config           'workerd',
  vite:config           'worker',
  vite:config           'astro',
  vite:config           'astro'
  vite:config         ]
  vite:config       },
  vite:config       target: 'webworker'
  vite:config     },
```

Note that the conditions lists browser before `workerd`. While it's [not well documented in vite docs](https://vite.dev/config/shared-options.html#resolve-conditions) it is my understanding that the order in this array matters, and conditions listed earlier will have higher precedence than conditions listed later.  Fortunately this doesn't seem to cause problems in this simple Astro repro application from @phase. But I think it would be good to change the code to use `unshift` instead of `push` to prepend `workerd` to the array, rather than append it.


So to sum up my resolution recommendation:
- [ ] remove the `alias` config
- [ ] remove the `externals` config
- [ ] (optional) prepend `workerd` to the `conditions` array rather than appending

**Comment by ematipico** [comment]
Can't git blame help explaining the reason why certain things were added? I believe the PRs associated with them can provide some useful context

**Comment by IgorMinar** [comment]
You are right @bluwy! In my local changes I did set the externals to [] but forgot to document it! I'll update my comment.

**Comment by bluwy** [comment]
I've sent a PR to remove React's `ssr.external` config (https://github.com/withastro/astro/pull/12996) and a PR to cleanup the Vue stuff in the Cloudflare integration (https://github.com/withastro/adapters/pull/506). I think what's left to fix the issue is to remove the react alias in the cloudflare integration and verify if it still works in React 18 and 19.

About forcing `ssr.external` to `[]`, I think we can perhaps discuss about doing that separately as I think there's value still in explicitly externalizing certain modules for some reason (maybe to workaround stuff?). But it could also be possible to the cloudflare integration to be more strict and warn about those explicitly set externals perhaps.

**Comment by IgorMinar** [comment]
Great work @bluwy! Thank you.

This should be enough to resolve the original issue.

As for supporting externals, I just don't see how that would work since workerd doesn't have access to node_modules. I suppose you could find a creative way to still take advantage of externals if you really knew what you were doing, but I find it surprising that the externals config has a higher precedence than noExternals which is set by the plugin. This seems like a trap waiting to waste someone's time in the future. But I'll leave it to you to decide what to do here.

Thank you for the prs!

**Comment by etc-tiago** [comment]
> here's a workaround [withastro/adapters#436 (comment)](https://github.com/withastro/adapters/pull/436#issuecomment-2525190557)
> 
> `astro.config.mjs`
> 
> // @ts-check
> import { defineConfig } from "astro/config";
> import cloudflare from "@astrojs/cloudflare";
> 
> import react from "@astrojs/react";
> 
> // https://astro.build/config
> export default defineConfig({
>   output: "server",
>   adapter: cloudflare(),
>   integrations: [react()],
>   vite: {
>     resolve: {
>       // Use react-dom/server.edge instead of react-dom/server.browser for React 19.
>       // Without this, MessageChannel from node:worker_threads needs to be polyfilled.
>       alias: import.meta.env.PROD && {
>         "react-dom/server": "react-dom/server.edge",
>       },
>     },
>   },
> });

It's work!

**Comment by ohroy** [comment]
wrangler preview still show this error

**Comment by leads** [comment]
> here's a workaround [withastro/adapters#436 (comment)](https://github.com/withastro/adapters/pull/436#issuecomment-2525190557)
> 
> `astro.config.mjs`
> 
> // @ts-check
> import { defineConfig } from "astro/config";
> import cloudflare from "@astrojs/cloudflare";
> 
> import react from "@astrojs/react";
> 
> // https://astro.build/config
> export default defineConfig({
>   output: "server",
>   adapter: cloudflare(),
>   integrations: [react()],
>   vite: {
>     resolve: {
>       // Use react-dom/server.edge instead of react-dom/server.browser for React 19.
>       // Without this, MessageChannel from node:worker_threads needs to be polyfilled.
>       alias: import.meta.env.PROD && {
>         "react-dom/server": "react-dom/server.edge",
>       },
>     },
>   },
> });

this works for me, using Astro
`"dependencies": {
    "@astrojs/cloudflare": "^12.2.1",
    "@astrojs/react": "^4.2.0",
    "astro": "^5.2.5",
    "react": "^19.0.0",
    "react-dom": "^19.0.0"
  }`

not sure if it should still be needed? My project is a minimal astro project hosted on Cloudflare Pages

**Comment by rptrainor** [comment]
This is still happening with a React app using the Astro Cloudflare adapter. Thank you for the workarounds, but the failed deployment is what brought me here

**Comment by pavi2410** [comment]
The issue still persists

**Comment by anthonycallo04** [comment]
Hi! Did you find a solution? I'm also having the same problem when using React 19 and trying to deploy to Cloudflare Pages, it shows me the error: Error: Failed to publish your Function. Got error: Uncaught ReferenceError: MessageChannel is not defined

**Comment by jgarrard** [comment]
Getting this same error MessageChannel is not defined. I'm not using React in my site, but I am using Keystatic, which I believe uses React for their components.

**Comment by m4rrc0** [comment]
> Getting this same error MessageChannel is not defined. I'm not using React in my site, but I am using Keystatic, which I believe uses React for their components.

@jgarrard , also came here a few days back while trying out Keystatic. The workaround in the [comment above](https://github.com/withastro/astro/issues/12824#issuecomment-2563095382) worked for me.

I created [this repo](https://github.com/m4rrc0/keystatic-deploy-test/) to test Keystatic (with Astro) deployments on the 3 hosting providers supported by Astro. This can probably be improved but all 3 deployments are working currently.  
The part that most interests you is [the vite config specific to Cloudflare](https://github.com/m4rrc0/keystatic-deploy-test/blob/251d537272eba69f95a5dd89aa801a2b21b72445/astro.config.mjs#L18). Also have a look at the `wrangler.jsonc` file and the [Cloudflare Options](https://github.com/m4rrc0/keystatic-deploy-test/blob/251d537272eba69f95a5dd89aa801a2b21b72445/astro.config.mjs#L10) I pass to the adapter.

**Comment by lgh06** [comment]
same issue.
It's so bad to revert my React version to 18 to run my astro site on cloudflare.

**Comment by mlafeldt** [comment]
Author from the [original issue](https://github.com/withastro/adapters/pull/436) here. 👋

I appreciate the ongoing work on the problem. However, how will fixing bundling change the fact that `MessageChannel` isn't available in the Workers runtime?

You can certainly polyfill it, e.g. via unenv, but using `react-dom/server.edge` as a renderer still seems to be the easiest fix (which I've been using in production for over half a year now), as it doesn't use `MessageChannel`. (Besides, the edge renderer is meant for stripped-down edge environments such as Cloudflare Workers, no?)

Unfortunately, this approach will break React 18, which is why my original fix was rejected. Again, please take a look at my findings in the original issue.

**Comment by mlafeldt** [comment]
> I appreciate the ongoing work on the problem. However, how will fixing bundling change the fact that `MessageChannel` isn't available in the Workers runtime?

I guess the goal would be to auto-select the edge renderer when targeting Cloudflare? That would be great, but we're not there yet.

**Comment by yuvadm** [comment]
FWIW, the type error in the workaround from https://github.com/withastro/astro/issues/12824#issuecomment-2563095382 can be fixed with:

```ts
resolve: {
  // Use react-dom/server.edge instead of react-dom/server.browser for React 19.
  // Without this, MessageChannel from node:worker_threads needs to be polyfilled.
  alias: import.meta.env.PROD ? {
    "react-dom/server": "react-dom/server.edge",
  } : undefined,
},
```

**Comment by alexanderniebuhr** [comment]
Just came across this. So we know the solution for React 19, but it is not compatible with React 18, so we need to figure out how to support both versions.

**Comment by ericclemmons** [comment]
Welp.  Hours later, _this_ was the fix (**both** `ssr.alias` and `resolve.alias`!):

```ts
import cloudflare from "@astrojs/cloudflare";
import react from "@astrojs/react";
import tailwindcss from "@tailwindcss/vite";
import { defineConfig } from "astro/config";

// https://github.com/withastro/astro/issues/12824
const alias = import.meta.env.PROD ? {
  "react-dom/server": "react-dom/server.edge",
} : undefined;

// https://astro.build/config
export default defineConfig({
  adapter: cloudflare(),
  integrations: [react()],
  vite: {
    plugins: [tailwindcss()],
    resolve: { alias },
    ssr: { alias },
  },
  output: "server",
});
```

**Comment by connebs** [comment]
> Just came across this. So we know the solution for React 19, but it is not compatible with React 18, so we need to figure out how to support both versions.

Why do you need to support multiple React versions in a single project?

**Comment by aifirstd3v** [comment]
> Welp. Hours later, _this_ was the fix (**both** `ssr.alias` and `resolve.alias`!):
> 
> import cloudflare from "@astrojs/cloudflare";
> import react from "@astrojs/react";
> import tailwindcss from "@tailwindcss/vite";
> import { defineConfig } from "astro/config";
> 
> // https://github.com/withastro/astro/issues/12824
> const alias = import.meta.env.PROD ? {
>   "react-dom/server": "react-dom/server.edge",
> } : undefined;
> 
> // https://astro.build/config
> export default defineConfig({
>   adapter: cloudflare(),
>   integrations: [react()],
>   vite: {
>     plugins: [tailwindcss()],
>     resolve: { alias },
>     ssr: { alias },
>   },
>   output: "server",
> });


`Object literal may only specify known properties, and 'alias' does not exist in type 'SSROptions'`
I am using Astro 5.11.

**Comment by rialco** [comment]
> Welp. Hours later, _this_ was the fix (**both** `ssr.alias` and `resolve.alias`!):
> 
> import cloudflare from "@astrojs/cloudflare";
> import react from "@astrojs/react";
> import tailwindcss from "@tailwindcss/vite";
> import { defineConfig } from "astro/config";
> 
> // https://github.com/withastro/astro/issues/12824
> const alias = import.meta.env.PROD ? {
>   "react-dom/server": "react-dom/server.edge",
> } : undefined;
> 
> // https://astro.build/config
> export default defineConfig({
>   adapter: cloudflare(),
>   integrations: [react()],
>   vite: {
>     plugins: [tailwindcss()],
>     resolve: { alias },
>     ssr: { alias },
>   },
>   output: "server",
> });

Hey thanks for your solution! I was wondering if I would be needing your approach or the first one proposed. Could you explain why you added the ssr portion?

edit: it returns `Type '{ "react-dom/server"?: string | undefined; }' has no properties in common with type 'SSROptions'.ts(2559)
` btw

**Comment by icepuma** [comment]
Will https://developers.cloudflare.com/changelog/2025-08-11-messagechannel/ help to fix this issue?

**Comment by devinpitcher** [comment]
> Will https://developers.cloudflare.com/changelog/2025-08-11-messagechannel/ help to fix this issue?

Setting `"compatibility_date": "2025-08-15"` worked for me without needing to edit my Astro/Vite config.

**Comment by quackerex** [comment]
> Setting `"compatibility_date": "2025-08-15"` worked for me without needing to edit my Astro/Vite config.

That didn't work for me :(

**Comment by ematipico** [comment]
Could you try Astro v6 beta, with the beta version of the cloudflare adapter? I believe this issue is fixed already

**Comment by matthewp** [comment]
I updated the reproduction and it seems to work now with Astro 6 beta: https://github.com/matthewp/test-astro-react19-cf-workers

Going to close, if anyone seems differently let me know and I'll reopen.
